# Automated detection and annotation of RNA virus subgenomic RNAs through analysis of public RNA-seq data

**Created during my 8-week, Summer Internship in the Firth Lab at the University of Cambridge, Department of Pathology, Division of Virology**

July 2024 - September 2024


Funded by: "Early Career Researcher Seed Fund Award" granted by the C2D3 Cambridge Centre for Data-Driven Discovery


Project Description: Many +ssRNA viruses, such as SARS-CoV-2, Zika virus, Dengue virus, etc., produce one or more shorter transcripts (alongside their main polyprotein), known as sub-genomic RNAs (sgRNAs), from which additional viral proteins are translated. These sgRNAs are essential to the pathogenicity of many viruses, including the coronaviruses and flaviviruses and are important targets in drug and vaccine development.

In order to understand and target viral sgRNAs, we first need to be able to identify them, however, current methods are time consuming and expensive. Publicly available RNA-seq transcriptomic datasets offer a way forward. Thus, the main aim is to develop robust computational methods to leverage publicly available high throughput sequencing data to systematically determine the sgRNA landscape of thousands of viral species. To achieve this, this project will involve the development of a computational pipeline, incorporating AI-based classification, to predict the sgRNAs for viruses found in high throughput RNA-sequencing (RNA-seq) datasets.

Overall, the development of this pipeline will expand our knowledge of the molecular biology of poorly studied virus groups. It will also improve genome annotations of better known virus groups (such as coronaviruses), leading to a better understanding of the pathogenesis of these viruses and potentially allowing for novel treatment options or vaccine targets. It will also expedite the identification of sgRNAs in newly identified viruses, increasing our preparedness for future viral outbreaks.